File: cli/narrator/skip.py
import re

_TOC_HEADINGS = {'table of contents', 'contents'}

_FRONT_MATTER_KEYWORDS = {
    'dedication',
    'dedicated to',
    'to my ',
    'for my ',
    'acknowledgements',
    'acknowledgments',
    'appreciation',
    'foreword',
    'preface',
    'copyright \xa9',
    'copyright (c)',
    'all rights reserved',
    'isbn',
    'published by',
    'first published',
    'printed in',
}

# Lines that look like "Chapter Title ............. 12" or "Chapter Title    12"
_TOC_LINE = re.compile(r'.{3,}(?:\.{3,}|\s{3,})\d+\s*$')


def detect_candidates(pages: list[str], limit: int = 20) -> list[int]:
    candidates: list[int] = []
    for i, text in enumerate(pages[:limit]):
        stripped = text.strip()
        if not stripped:
            candidates.append(i)
        elif _is_toc(stripped) or _is_front_matter(stripped):
            candidates.append(i)
    return candidates


def _is_toc(text: str) -> bool:
    lines = text.split('\n')
    lower_first = [l.lower().strip() for l in lines[:6]]

    if any(l in _TOC_HEADINGS for l in lower_first):
        return True

    toc_count = sum(1 for l in lines if _TOC_LINE.match(l.strip()))
    if lines and toc_count / len(lines) > 0.35:
        return True

    return False


def _is_front_matter(text: str) -> bool:
    word_count = len(text.split())
    if word_count > 350:
        return False
    lower = text.lower()
    return any(kw in lower for kw in _FRONT_MATTER_KEYWORDS)

File: pwa/src/lib/skip.ts
const TOC_HEADINGS = new Set(['table of contents', 'contents']);

const FRONT_MATTER_KEYWORDS = [
  'dedication',
  'dedicated to',
  'to my ',
  'for my ',
  'acknowledgements',
  'acknowledgments',
  'appreciation',
  'foreword',
  'preface',
  'copyright ©',
  'copyright (c)',
  'all rights reserved',
  'isbn',
  'published by',
  'first published',
  'printed in',
  'about the publisher',
  'about the author',
  'also by ',
  'other books by',
  'by the same author',
  'transcribed by',
  'volume one',
  'volume two',
  'volume three',
  'volume i ',
  'volume ii ',
  'volume iii ',
];

const TOC_LINE = /^.{3,}(?:\.{3,}|\s{3,})\d+\s*$/;

export function detectCandidates(pages: string[], limit = 20): number[] {
  const candidates: number[] = [];
  for (let i = 0; i < Math.min(pages.length, limit); i++) {
    const stripped = pages[i].trim();
    if (!stripped || isToc(stripped) || isFrontMatter(stripped)) {
      candidates.push(i);
    }
  }
  return candidates;
}

function isToc(text: string): boolean {
  const lines = text.split('\n');
  const lowerFirst = lines.slice(0, 6).map((l) => l.toLowerCase().trim());

  if (lowerFirst.some((l) => TOC_HEADINGS.has(l))) return true;

  const tocCount = lines.filter((l) => TOC_LINE.test(l.trim())).length;
  if (lines.length > 0 && tocCount / lines.length > 0.35) return true;

  return false;
}

function isFrontMatter(text: string): boolean {
  const wordCount = text.split(/\s+/).length;
  if (wordCount > 350) return false;
  // Very short pages are title cards, publisher stamps, or filler — always skip
  if (wordCount < 15) return true;
  const lower = text.toLowerCase();
  return FRONT_MATTER_KEYWORDS.some((kw) => lower.includes(kw));
}
